### PR TITLE
Add 3D keypoints output to TFJS blazepose

### PR DESCRIPTION
**Pull request by ahmedsabie**
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/801)

### CODE REVIEW

**Review by lina128 — COMMENTED** [review]
Great work, very neat, thank you Ahmed! LGTM overall, just some minor comments, see below.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-:-MgrYq4122zijnvB9Hu9:b-eims2c)* status: 0 of 1 approvals obtained (waiting on @ahmedsabie)

___
*[pose-detection/src/blazepose_tfjs/constants.ts, line 25 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrLuBm316BwdbBS-ga:-MgrLuBm316BwdbBS-gb:b-bomdxt) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/constants.ts#L25)):*
> ```
> heavy
> ```
lite?
___
*[pose-detection/src/blazepose_tfjs/constants.ts, line 27 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrLw19EhCZ2c7zdrdv:-MgrLw19EhCZ2c7zdrdw:b-65082b) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/constants.ts#L27)):*
> ```
> lite
> ```
heavy?
___
*[pose-detection/src/blazepose_tfjs/constants.ts, line 27 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrM0veAiSQXzT6QA4U:-MgrM0veAiSQXzT6QA4V:borwx3i) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/constants.ts#L27)):*
> ```
> blazepose_ghum_39kp_lite_2021_05_06_named
> ```
It's better not exposing these information (such as date) to the public. Can we rename the url?
___
*[pose-detection/src/blazepose_tfjs/constants.ts, line 153 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrR90hADC5Y0eLHn1-:-MgrR90hADC5Y0eLHn10:b-cos4qi) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/constants.ts#L153)):*
> ```
> disableValueScaling: true
> ```
Add this annotation: As world landmdarks are predicted in real world 3D coordintates in meters (rather than in pixels of input image) prediction scale does not depend on the pose size in the image.
___
*[pose-detection/src/blazepose_tfjs/detector.ts, line 334 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrSgtz0lSQNgrBOkt6:-MgrSgtz0lSQNgrBOkt7:bk55cyv) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/detector.ts#L334)):*
> ```
> score.
> ```
reformat this line.
___
*[pose-detection/src/blazepose_tfjs/detector.ts, line 336 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrSkuz5w8Xhj47ije_:-MgrSkuz5w8Xhj47ijea:b2euw7s) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/detector.ts#L336)):*
> ```
> heatmap for the 39 landmarks.
> ```
reformat this line.
___
*[pose-detection/src/blazepose_tfjs/detector.ts, line 403 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrUJOSCEv_n4lc8i49:-MgrUJOSCEv_n4lc8i4A:bna6q8r) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/detector.ts#L403)):*
> ```
>     // Decodes the world landmark tensors into a list of landmarks, where the
>     // landmark coordinates are normalized by the size of the input image to
>     // the model.
> ```
Just this part, because the world landmarks are not normalized by the size of the image:
// Decodes the world landmark tensors into a list of landmarks.
___
*[pose-detection/src/calculators/keypoints_one_euro_filter.ts, line 54 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrXAB38LmF_YiHLtoL:-MgrXAB4DZKZHBQYw62B:bybgb1x) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/calculators/keypoints_one_euro_filter.ts#L54)):*
> ```
> disableValueScaling
> ```
Given the behavior change, can you also change the configs for MoveNet?
Add `disableValueScaling: true` here:
https://github.com/tensorflow/tfjs-models/blob/master/pose-detection/src/movenet/constants.ts#L51

**Review by ahmedsabie — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/801)* status: 0 of 1 approvals obtained (waiting on @ivan-grishchenko and @lina128)

___
*[pose-detection/src/blazepose_tfjs/constants.ts, line 25 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrLuBm316BwdbBS-ga:-Mgryko56ebvRrBFrFfb:b-gj8tya) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/constants.ts#L25)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> heavy
> ```
lite?
</blockquote></details>

Great catch!
___
*[pose-detection/src/blazepose_tfjs/constants.ts, line 27 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrM0veAiSQXzT6QA4U:-Mgrymcx6uG7KsM3P1Ge:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/constants.ts#L27)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> blazepose_ghum_39kp_lite_2021_05_06_named
> ```
It's better not exposing these information (such as date) to the public. Can we rename the url?
</blockquote></details>

Done.
___
*[pose-detection/src/blazepose_tfjs/constants.ts, line 27 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrLw19EhCZ2c7zdrdv:-MgryoCz49XpuP2AYx16:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/constants.ts#L27)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> lite
> ```
heavy?
</blockquote></details>

Done.
___
*[pose-detection/src/blazepose_tfjs/constants.ts, line 153 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrR90hADC5Y0eLHn1-:-Mgryn1t0PvywKGX02Gh:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/constants.ts#L153)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> disableValueScaling: true
> ```
Add this annotation: As world landmdarks are predicted in real world 3D coordintates in meters (rather than in pixels of input image) prediction scale does not depend on the pose size in the image.
</blockquote></details>

Done.
___
*[pose-detection/src/blazepose_tfjs/detector.ts, line 334 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrSgtz0lSQNgrBOkt6:-MgryfMKEtSaCzcJ7qbj:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/detector.ts#L334)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> score.
> ```
reformat this line.
</blockquote></details>

Done.
___
*[pose-detection/src/blazepose_tfjs/detector.ts, line 336 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrSkuz5w8Xhj47ije_:-MgryfhR3yC-NsnmLUES:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/detector.ts#L336)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> heatmap for the 39 landmarks.
> ```
reformat this line.
</blockquote></details>

Done.
___
*[pose-detection/src/blazepose_tfjs/detector.ts, line 403 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrUJOSCEv_n4lc8i49:-MgrygLJ5z8yVpAXBwCX:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/blazepose_tfjs/detector.ts#L403)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
>     // Decodes the world landmark tensors into a list of landmarks, where the
>     // landmark coordinates are normalized by the size of the input image to
>     // the model.
> ```
Just this part, because the world landmarks are not normalized by the size of the image:
// Decodes the world landmark tensors into a list of landmarks.
</blockquote></details>

Done.
___
*[pose-detection/src/calculators/keypoints_one_euro_filter.ts, line 54 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrXAB38LmF_YiHLtoL:-MgryhPN-CdyGQyAMtUi:b-896fix) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/calculators/keypoints_one_euro_filter.ts#L54)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> disableValueScaling
> ```
Given the behavior change, can you also change the configs for MoveNet?
Add `disableValueScaling: true` here:
https://github.com/tensorflow/tfjs-models/blob/master/pose-detection/src/movenet/constants.ts#L51

</blockquote></details>

Done.

**Review by lina128 — APPROVED** [review]
LGTM from TFJS, also wait for Ivan’s approval.

File: pose-detection/src/calculators/get_object_scale.ts
Comment (ivan-grishchenko) [inline pose-detection/src/calculators/get_object_scale.ts]: This commit doesn't look like support of non-normalized keypoints.
Do you just remove scaling in the file here?
If so - I don't see it reflected in the code that actually calls this and other methods.

File: pose-detection/src/calculators/keypoints_one_euro_filter.ts
Comment (ivan-grishchenko) [inline pose-detection/src/calculators/keypoints_one_euro_filter.ts:53]: Why is this logic removed? It's mandatory for BlazePose smoothing.
Same for all changes below.

**Review by ivan-grishchenko — CHANGES_REQUESTED** [review]

File: pose-detection/src/blazepose_tfjs/calculators/calculate_world_landmark_projection.ts
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose_tfjs/calculators/calculate_world_landmark_projection.ts:29]: I would rather name this parameter "world_landmarks".

**Review by ivan-grishchenko — CHANGES_REQUESTED** [review]

File: pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts:26]: Should it be Keypoint type?
Or Keypoint is a parent class for LandmarkList and NormalizedLandmarklist?

File: pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts:41]: Please mention this parameter in documentation

File: pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts:42]: As the field is named "score" instead of "visibility" please name the class, method and parameter accordingly.

**Review by ivan-grishchenko — CHANGES_REQUESTED** [review]

File: pose-detection/src/blazepose_tfjs/calculators/calculate_landmark_projection.ts
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose_tfjs/calculators/calculate_landmark_projection.ts]: I checked the logic, but consider adding test cases from https://github.com/google/mediapipe/blob/master/mediapipe/calculators/util/landmark_projection_calculator_test.cc

File: pose-detection/src/blazepose_tfjs/detector.ts
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose_tfjs/detector.ts]: Why output indexes are different in lite/full/heavy?

File: pose-detection/src/blazepose_tfjs/detector.ts
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose_tfjs/detector.ts:403]: World landmarks shouldn't be normalized by the image size and your comment states that they are (although not sure what is happening inside the method). Please update comment and check that nothing happens to world landmarks in the method (i.e. returned as they are from the model).

File: pose-detection/src/blazepose_tfjs/detector.ts
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose_tfjs/detector.ts]: Two things here:
- How "normalized = true" affects smoothing? Will it be scaled with image size? It yes, then set to false and pass `null` as imageSize to be safe.
- I know that world landmarks smoothing config says not to scale, but to be safe pass either `null` or `1` as a scale and add comment about it.

File: pose-detection/src/calculators/keypoints_smoothing.ts
Comment (ivan-grishchenko) [inline pose-detection/src/calculators/keypoints_smoothing.ts]: Looks like it changes logic for regular landmarks as well right?
If true - move to separate CL, cause it's a pretty significant change and is lost among other changes in this CL.

File: pose-detection/src/calculators/keypoints_one_euro_filter.ts
Comment (ivan-grishchenko) [inline pose-detection/src/calculators/keypoints_one_euro_filter.ts:53]: Why you remove this logic in first CL in the chain and then revert it here?
Same for all changes below.

**Review by ivan-grishchenko — CHANGES_REQUESTED** [review]

File: pose-detection/src/calculators/interfaces/config_interfaces.ts
Comment (ivan-grishchenko) [inline pose-detection/src/calculators/interfaces/config_interfaces.ts:67]: Added but not used

File: pose-detection/src/calculators/keypoints_one_euro_filter.ts
Comment (ivan-grishchenko) [inline pose-detection/src/calculators/keypoints_one_euro_filter.ts]: disableValueScaling should be used here

**Review by ivan-grishchenko — CHANGES_REQUESTED** [review]

File: pose-detection/src/calculators/keypoints_one_euro_filter.ts
Comment (ivan-grishchenko) [inline pose-detection/src/calculators/keypoints_one_euro_filter.ts:54]: That should be in previous CL

File: pose-detection/src/blazepose_tfjs/calculators/calculate_landmark_projection.ts
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose_tfjs/calculators/calculate_landmark_projection.ts]: I don't see CL where it was removed. It was still there in "Add WorldLandmarkProjectionCalculator" CL.
And why it was removed in a first place?

**Review by ivan-grishchenko — CHANGES_REQUESTED** [review]

**Comment by ahmedsabie** [comment]
Hey @ivan-grishchenko 

Sorry im a bit confused by the comments, are you reviewing commit by commit? The commits weren't meant to be valid on their own as I was initially unfamiliar with the code, so I was exploring and making changes temporarily, can you take a look only at the cumulative changes

**Comment by ahmedsabie** [comment]
I fixed the comments relating to the cumulative changes

**Review by lina128 — COMMENTED** [review]
Hi @ivan-grishchenko, thank you for the detailed review, we will fix the commits and address your review.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-:-MhEjggI0RC5PLsGvZNn:b-isrnzp)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128)

**Comment by ahmedsabie** [comment]
@ivan-grishchenko 
Just fixed up the commits, when you get a chance let me know if everything is OK thanks!

**Review by lina128 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/801)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @ahmedsabie and @lina128)

___
*[pose-detection/src/calculators/keypoints_one_euro_filter.ts, line 54 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrXAB38LmF_YiHLtoL:-MhJxvqd1un-SIRqL-AC:bocb2l1) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/calculators/keypoints_one_euro_filter.ts#L54)):*
<details><summary><i>Previously, ahmedsabie wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I don't see this change.

File: pose-detection/src/calculators/get_object_scale.ts
Comment (ahmedsabie) [reply pose-detection/src/calculators/get_object_scale.ts]: This change was temporary and is now gone in the newer commits

File: pose-detection/src/calculators/keypoints_one_euro_filter.ts
Comment (ahmedsabie) [reply pose-detection/src/calculators/keypoints_one_euro_filter.ts:53]: This change was temporary and is now gone in the newer commits

File: pose-detection/src/blazepose_tfjs/calculators/calculate_world_landmark_projection.ts
Comment (ahmedsabie) [reply pose-detection/src/blazepose_tfjs/calculators/calculate_world_landmark_projection.ts]: Fixed in commit Rename variables

File: pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts
Comment (ahmedsabie) [reply pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts]: Keypoint type, fixed in Rename variables commit

File: pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts
Comment (ahmedsabie) [reply pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts]: Fixed in Rename variables commit

File: pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts
Comment (ahmedsabie) [reply pose-detection/src/blazepose_tfjs/calculators/calculate_visibility_copy.ts]: Fixed in Rename variables commit

File: pose-detection/src/blazepose_tfjs/detector.ts
Comment (ahmedsabie) [reply pose-detection/src/blazepose_tfjs/detector.ts]: The old converter we used to turn the model into TFJS didn't keep the output name and it didn't retain the outputs in a fixed order each time so this had to be hardcoded in. The new converter fixes that with named outputs and the update is found in the Update model to named outputs commit

File: pose-detection/src/blazepose_tfjs/detector.ts
Comment (ahmedsabie) [reply pose-detection/src/blazepose_tfjs/detector.ts:403]: This change was temporary and is now gone in the newer commits

File: pose-detection/src/blazepose_tfjs/detector.ts
Comment (ahmedsabie) [reply pose-detection/src/blazepose_tfjs/detector.ts]: This change was temporary and is now gone in the newer commits, it is now normalized = false

File: pose-detection/src/calculators/keypoints_smoothing.ts
Comment (ahmedsabie) [reply pose-detection/src/calculators/keypoints_smoothing.ts]: This change was temporary and is now gone in the newer commits

File: pose-detection/src/calculators/keypoints_one_euro_filter.ts
Comment (ahmedsabie) [reply pose-detection/src/calculators/keypoints_one_euro_filter.ts:53]: This change was temporary and is now gone in the newer commits

File: pose-detection/src/calculators/interfaces/config_interfaces.ts
Comment (ahmedsabie) [reply pose-detection/src/calculators/interfaces/config_interfaces.ts:67]: This change was temporary and is now added and used in  Add 3D keypoints output to TFJS blazepose commit

File: pose-detection/src/calculators/keypoints_one_euro_filter.ts
Comment (ahmedsabie) [reply pose-detection/src/calculators/keypoints_one_euro_filter.ts]: This change was temporary and is now added and used in Add 3D keypoints output to TFJS blazepose commit

File: pose-detection/src/calculators/keypoints_one_euro_filter.ts
Comment (ahmedsabie) [reply pose-detection/src/calculators/keypoints_one_euro_filter.ts:54]: This change was temporary and is now added and used in Add 3D keypoints output to TFJS blazepose commit

File: pose-detection/src/blazepose_tfjs/calculators/calculate_landmark_projection.ts
Comment (ahmedsabie) [reply pose-detection/src/blazepose_tfjs/calculators/calculate_landmark_projection.ts]: This change was temporary and is now gone in the newer commits

File: pose-detection/src/blazepose_tfjs/calculators/calculate_landmark_projection.ts
Comment (ahmedsabie) [reply pose-detection/src/blazepose_tfjs/calculators/calculate_landmark_projection.ts]: Added in Add landmark projection test commit

**Review by ahmedsabie — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-models/801)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128)

___
*[pose-detection/src/calculators/keypoints_one_euro_filter.ts, line 54 at r2](https://reviewable.io/reviews/tensorflow/tfjs-models/801#-MgrXAB38LmF_YiHLtoL:-MhKC2P91Ek8q0HPAFqt:b-vjx2m1) ([raw file](https://github.com/tensorflow/tfjs-models/blob/cb52c79b27ac99af9feaae7236058aaf63651384/pose-detection/src/calculators/keypoints_one_euro_filter.ts#L54)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

I don't see this change.
</blockquote></details>

It's found in this commit: https://github.com/tensorflow/tfjs-models/pull/801/commits/786f2c9d920e98667a61f0a0dc695319aa4af76c

**Review by ivan-grishchenko — APPROVED** [review]

File: pose-detection/src/blazepose_tfjs/detector.ts
Comment (ivan-grishchenko) [inline pose-detection/src/blazepose_tfjs/detector.ts:528]: Please use NULL (or whatever is used in JS) instead of imageSize and objectScaleROI to avoid bugs as world landmarks filtering doesn't need those parameters.

**Review by ivan-grishchenko — APPROVED** [review]

**Pull request merged**
